src/common/wrap/wrap-websocket.js

```diff
@@ -3,59 +3,167 @@
+    static toString () { // fake native WebSocket when static class is stringified
+      return 'function WebSocket() { [native code] }'
+    }
+
+    toString () { // fake [object WebSocket] when instance is stringified
+      return '[object WebSocket]'
+    }
+
+    get [Symbol.toStringTag] () { // fake [object WebSocket] when Object.prototype.toString.call is used on instance
+      return WrappedWebSocket.name
+    }
+
+    // Private method to tag send, close, and event listener errors with WebSocket ID for JSErrors feature
+    #tagError (error) {
+      ;(error.__newrelic ??= {}).socketId = this.nrData.socketId
+      this.nrData.hasErrors ??= true
+    }
 
     constructor (...args) {
       super(...args)
-      const socketId = generateRandomHexString(6)
-      this.report = reporter(socketId)
-      this.report('new')
-
-      const events = ['message', 'error', 'open', 'close']
-      /** add event listeners */
-      events.forEach(evt => {
-        this.addEventListener(evt, function (e) {
-          this.report(ADD_EVENT_LISTENER_TAG, { eventType: evt, event: e })
+      this.nrData = {
+        timestamp: now(), // this will be time corrected later when timeKeeper is avail
+        currentUrl: window.location.href,
+        socketId: generateRandomHexString(8),
+        requestedUrl: args[0],
+        requestedProtocols: Array.isArray(args[1]) ? args[1].join(',') : (args[1] || '')
+        // pageUrl will be set by addEvent later; unlike timestamp and currentUrl, it's not sensitive to *when* it is set
+      }
+
+      this.addEventListener('open', () => {
+        this.nrData.openedAt = now()
+        ;['protocol', 'extensions', 'binaryType'].forEach(prop => {
+          this.nrData[prop] = this[prop]
         })
       })
+
+      this.addEventListener('message', (event) => {
+        const { type, size } = getDataInfo(event.data)
+        this.nrData.messageOrigin ??= event.origin // the origin of messages thru WS lifetime cannot be changed, so set once is sufficient
+        this.nrData.messageCount = (this.nrData.messageCount ?? 0) + 1
+        this.nrData.messageBytes = (this.nrData.messageBytes ?? 0) + size
+        this.nrData.messageBytesMin = Math.min(this.nrData.messageBytesMin ?? Infinity, size)
+        this.nrData.messageBytesMax = Math.max(this.nrData.messageBytesMax ?? 0, size)
+        if (!(this.nrData.messageTypes ?? '').includes(type)) {
+          this.nrData.messageTypes = this.nrData.messageTypes ? `${this.nrData.messageTypes},${type}` : type
+        }
+      })
+
+      this.addEventListener('close', (event) => {
+        this.nrData.closedAt = now()
+        this.nrData.closeCode = event.code
+        this.nrData.closeReason = event.reason
+        this.nrData.closeWasClean = event.wasClean
+        this.nrData.connectedDuration = this.nrData.closedAt - this.nrData.openedAt
+
+        subscribedFeatures.forEach(featureName => handle('ws-complete', [this.nrData], this, featureName, wsEE))
```

Comment:
Is there a reason to break the emit pattern here that the other wrappers use in favor of directly using handle? (All the other wrappers "emit" an unscoped message and the instrument modules are responsible for the "handle" behavior to buffer for the agg modules).

Reply:
i understand, but this is another case of introducing "this feature works this way while this feature works that way" instead of a consistent implementation pattern, which we should avoid doing if we can help it.

Reply:
Ok but just know that creating extraneous execution loops because of an old indirection system is simply inefficient.
Changing

Reply:
extracted this convo out to a separate standalone ticket https://new-relic.atlassian.net/browse/NR-484610